### PR TITLE
Add a new dvstring facet method which is good for high-cardinality st…

### PR DESCRIPTION
**Pull request by timatbw**
…ring fields

### CODE REVIEW

**Comment by timatbw** [comment]
The reason for this facet method is because dvhash is not great when the field has high cardinality, regardless of the size of your result set.. how dvhash works is that it first builds up a large data structure mapping each distinct string value to a global ordinal, before processing the results and bucketing by global ordinal. If the field has millions of distinct values, this takes 10s or more, even if your search only has a small number of results.

So this dvstring alternative method works by using a Java HashMap with each string key being field values, then it works through the documents in each index segment, looking up the per-segment docvalues and finding the string key in our Map. This means the performance is more related to your result size, even if the field has huge cardinality.

I've tested this on a devserver, in both single and distributed searches, with sorting buckets by a stat or by count, and with allBuckets on or off. It works with subfacets and stats just like other methods. I've compared the responses for dvhash and dvstring and get the same output.

File: solr/core/src/java/org/apache/solr/search/facet/FacetFieldProcessorByHashDVString.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  }
+
+  private SimpleOrderedMap<Object> calcFacets() throws IOException {
+
+    int possibleValues = fcontext.base.size();
+    int hashSize = BitUtil.nextHighestPowerOfTwo((int) (possibleValues * (1 / 0.7) + 1));
+    hashSize = Math.min(hashSize, 1024);
+
+    table = new HashMap<>(hashSize);
+    slotList = new ArrayList<>();
```
Comment (johnmbw) [inline solr/core/src/java/org/apache/solr/search/facet/FacetFieldProcessorByHashDVString.java:92]: Does this needs presizing in any way too?

**Comment by johnmbw** [comment]
Not sure why the docker build is failing though.  @richardbadman maybe one for you to check?  It's complaining about:
```
#4 ERROR: docker.io/library/openjdk:8-stretch not found
```
For the `base.Dockerfile`, but from what I can tell that does exist...

File: solr/core/src/java/org/apache/solr/search/facet/FacetFieldProcessorByHashDVString.java (hunk 1 above)
Comment (timatbw) [reply solr/core/src/java/org/apache/solr/search/facet/FacetFieldProcessorByHashDVString.java:92]: It doesn't need to, really. And actually the hashmap size doesn't need to be picked either.. largely this code was started by copying the DVhash class and ripping out their custom hashtable. I could just remove this hashSize calc.

**Review by johnmbw — APPROVED** [review]

**Pull request merged**